sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -932,36 +1036,158 @@ public void setup() throws Exception {
           } else if (spec.getTopic() != null) {
             this.destination = session.createTopic(spec.getTopic());
           }
-
-          this.producer = this.session.createProducer(null);
+          this.producer = this.session.createProducer(this.destination);
+          this.isProducerNeedsToBeCreated = false;
         }
       }
 
-      @ProcessElement
-      public void processElement(ProcessContext ctx) {
+      public void publishMessage(T input) throws JMSException, JmsIOException {
         Destination destinationToSendTo = destination;
         try {
-          Message message = spec.getValueMapper().apply(ctx.element(), session);
+          Message message = spec.getValueMapper().apply(input, session);
           if (spec.getTopicNameMapper() != null) {
-            destinationToSendTo =
-                session.createTopic(spec.getTopicNameMapper().apply(ctx.element()));
+            destinationToSendTo = session.createTopic(spec.getTopicNameMapper().apply(input));
           }
           producer.send(destinationToSendTo, message);
-        } catch (Exception ex) {
-          LOG.error("Error sending message on topic {}", destinationToSendTo);
-          ctx.output(failedMessageTag, ctx.element());
+        } catch (JMSException | JmsIOException | NullPointerException exception) {
+          // Handle NPE in case of getValueMapper or getTopicNameMapper returns NPE
+          if (exception instanceof NullPointerException) {
+            throw new JmsIOException("An error occurred", exception);
+          }
+          throw exception;
+        }
+      }
+
+      public void close() throws JMSException {
+        isProducerNeedsToBeCreated = true;
+        if (producer != null) {
+          producer.close();
+          producer = null;
+        }
+        if (session != null) {
+          session.close();
+          session = null;
         }
+        if (connection != null) {
+          try {
+            // If the connection failed, stopping the connection will throw a JMSException
+            connection.stop();
+          } catch (JMSException exception) {
+            LOG.warn("The connection couldn't be closed", exception);
+          }
+          connection.close();
+          connection = null;
+        }
+      }
+    }
+
+    private static class JmsIOProducerFn<T> extends DoFn<T, T> {
```

Comment:
If I understand correctly, if should just work using JmsIOProduceRetryFn alone, why need another `JmsIOProducerFn` before it?

I personally think two step retry-write-on-failed is not a good pattern because it is unnecessarily complicated and more important it is hard to test (though I see it tested in unit test). It also requires to maintain two classes doing the same thing and keep them consistent for any future addition which is errorprone.

Reply:
I separated them because I'm not sure if the Backoff will hold the DoFn thread and I don't know if it will do that for the whole bundle (Which means the whole bundle will be in wait until the failed message is published). Just to not have any performances issues doing that. Do you think it would be better to keep in one DoFn ?

>PS: I tried to have a sort of cache of failed messages and flush them in Finishbundle cycle, but when throwing an exception in ProcessElement, the worker will call TearDown without FinishBundle.

Reply:
For messages go to JmsIOProduceRetryFn and have no error, if does not trigger backoff and do not hold DoFn thread.

Currently, if error happens, JmsIOProducerFn send failedMessage to JmsIOProduceRetryFn, if failed again it still triggers backoff. In runner JmsIOProducerFn and JmsIOProduceRetryFn will be fused and run on same thread. So if this is a concern it exists currently also

Reply:
Ok, so what's the best approach to keep the performance as it was ? should I just merge them into one ?

Reply:
For now I would simply use `JmsIOProduceRetryFn`. For succeeded send, the while (true) loop just breaks before checking backOff, it should has same performance as `JmsIOProducerFn`.

There are of course way of improvement. Currently I understand the `.send()` call is synchroneous (otherwise backoff won't work). Switch to cached, asynchronous API call would be more performant. An example in Beam's code base is KafkaWriter: https://github.com/apache/beam/blob/master/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java

Reply:
@Abacn Maybe we can for now merge the current solution, I will try to find time to make it better. I will keep in mind the example of kafka writer. I pushed a new commit with your recommendations.